src/ligero/utils.rs

```diff
@@ -0,0 +1,78 @@
+use ark_ff::PrimeField;
+
+use rayon::{iter::{ParallelIterator, IntoParallelRefIterator}, prelude::IndexedParallelIterator};
+
+pub(crate) struct Matrix<F: PrimeField> {
+    n: usize,
+    m: usize,
+    entries: Vec<Vec<F>>
+}
+
+impl<F: PrimeField> Matrix<F> {
+    
+    /// Returns a Matrix of dimensions n x m given a list of n * m field elements.
+    /// The list should be ordered row-first, i.e. [a11, ..., a1m, a21, ..., a2m, ...].
+    ///
+    /// # Panics
+    /// Panics if the dimensions do not match the length of the list
+    pub(crate) fn new_from_flat(n: usize, m: usize, entry_list: &[F]) -> Self {
+        assert_eq!(entry_list.len(), n * m, "Invalid matrix construction: dimensions are {} x {} but entry vector has {} entries", m, n, entry_list.len());
+
+        // TODO more efficient to run linearly?
+        let entries: Vec<Vec<F>> = (0..n).map(|row| (0..m).map(|col| entry_list[n * row + m]).collect()).collect();
+
+        Self {n, m, entries}
+    }
+
+    /// Returns a Matrix given a list of its rows, each in turn represented as a list of field elements.
+    ///
+    /// # Panics
+    /// Panics if the sub-lists do not all have the same length.
+    pub(crate) fn new_from_row_list(n: usize, m: usize, row_list: Vec<Vec<F>>) -> Self {
+        let mut m = row_list[0].len();
+
+        for row in row_list.iter() {
+            assert_eq!(row.len(), m, "Invalid matrix construction: not all rows have the same length");
+            m = row.len()
```

Comment:
that is not needed, `m` need not be mutable, since it will never change with the assert in place